plugin/qbackendconnection.cpp

```diff
@@ -391,44 +381,144 @@ void QBackendConnection::handleMessage(const QByteArray &message)
+        doDeliver = false;
+    }
+
+    // If the connection is not yet established, then block the message until
+    // callbacks have been installed.
+    if (m_state != ConnectionState::Established && !m_syncCallbacks.size()) {
+        qCDebug(lcConnection) << "Queueing handling of " << command << " as the connection is not yet established, and no callbacks have been installed";
+        doDeliver = false;
+    }
+
+    // Check the command against the state. Don't allow a command to deliver
+    // unless it's in the right state!
+    switch (m_state) {
+    case ConnectionState::WantVersion:
+        if (command != "VERSION") {
+            qCDebug(lcConnection) << "Queueing handling of " << command << " as we are WantVersion";
+            doDeliver = false;
+        }
+        break;
+    case ConnectionState::WantTypes:
+        if (command != "CREATABLE_TYPES") {
+            qCDebug(lcConnection) << "Queueing handling of " << command << " as we are WantTypes";
+            doDeliver = false;
+        }
+        break;
+    case ConnectionState::WantEngine:
+        // Need a QML engine pointer. Queue everything except further ROOT.
+        qCDebug(lcConnection) << "Queueing handling of " << command << " as we WantEngine";
+        doDeliver = false;
+        break;
+    case ConnectionState::WantRoot:
+        if (command != "ROOT") {
+            qCDebug(lcConnection) << "Queueing handling of " << command << " as we are WantRoot";
+            doDeliver = false;
         }
+        break;
+    case ConnectionState::Established:
+        break; // no queueing based on state.
+    }
+
+    if (!doDeliver) {
         m_pendingMessages.append(cmd);
         return;
     }
 
-    if (m_syncCallback) {
-        // Reset callback, set result, and handle immediately, bypassing the queue
+    // Pass the message to all sync callbacks to see if any of them want it.
+    // If so, handle it immediately, bypassing the queue.
+    bool wasHandled = false;
+    for (int i = 0; i < m_syncCallbacks.size(); i++) {
+        // Remove callback, set result, and handle immediately, bypassing the queue
         // even if out of order; see waitForMessage for details.
-        m_syncCallback = nullptr;
-        m_syncResult = cmd;
-    } else if (!m_pendingMessages.isEmpty()) {
-        // Other messages are pending, so queue this one as well
-        qCDebug(lcConnection) << "Queuing handling of" << command << "into non-empty message queue";
-        // In this case, there's no need to trigger the handler later; something already has
-        m_pendingMessages.append(cmd);
-        return;
+        if (m_syncCallbacks.at(i)(cmd)) {
+            m_syncCallbacks.removeAt(i);
+            m_syncResult = cmd;
+            wasHandled = true;
+            break;
+        }
+    }
+
+    if (!m_syncCallbacks.isEmpty() && !wasHandled) {
+        if (command == "ROOT" && m_state == ConnectionState::WantRoot) {
```

Comment:
This is the one part I'm really uncertain about. I've seen scenarios where there was never a callback for ROOT installed, so blocking continued endlessly. I think it looks like this might happen if the Backend singleton is not used.

Perhaps there's a different way to address it, though.

Reply:
Ah, yes! There was/is a bug here.

The relevant facts are:
 * Because ROOT contains values, it cannot be handled until there is an engine.
 * The engine is set on the first reference to `Backend` _or_ the first created instantiable type. Nothing else can proceed until this happens.
 * ROOT will always be immediately after CREATEABLE_TYPES (and therefore the third message on the connection).

This means ROOT ties in nicely to the state machine, but we might not be synchronously waiting on it. The connection is always:
 1. VERSION
 2. CREATEABLE_TYPES
 3. Queue _all_ messages until a QML engine is attached. It is not possible for any sync calls to be waiting.
 4. A QML engine is attached from a reference to `Backend` or through instantiation
 5. The next message is always ROOT, which is now possible to handle
 6. Everything is fully established, and it's possible for other sync messages to exist

In fewer words, I think the correct behavior is to pause and queue everything until there is a qml engine, then handle ROOT (unblocking its sync handler if present), and only then should any other messages -- including any sync messages other than ROOT -- be handled.

This won't work as you have it because one of the conditions above unsets `doDeliver` if there's no sync callbacks in the WantRoot state. I think the states there could be made cleaner with this all in mind, though..

Reply:
I realized that this sort of conflicts with my LIFO sync messages advice below, and then realized that it's much more simple than I thought.

ROOT is always sent, but it isn't actually _used_ until something calls `rootObject()` (such as using `Backend`). It doesn't actually need special treatment: if something wants it, `ensureRootObject()` gets called and will `waitForMessage`. There's no reason to process it early.

So I think the connection is "ready" as soon as there's a QML engine, because at that point it's able to handle any sort of message. ROOT can come through just like anything else.